src/lava/utils/weightutils.py

```diff
@@ -17,81 +18,154 @@ class SignMode(Enum):
-                         loihi2: bool = False) -> ty.Tuple[np.ndarray,
-                                                           int,
-                                                           int,
-                                                           SignMode]:
+def determine_sign_mode(weights: np.ndarray) -> SignMode:
+    if np.max(weights) < 0:
+        sign_mode = SignMode.INHIBITORY
+    elif np.min(weights) >= 0:
+        sign_mode = SignMode.EXCITATORY
+    else:
+        sign_mode = SignMode.MIXED
+
+    return sign_mode
+
+
+@dataclass
+class OptimizedWeights:
+    weights: np.ndarray
+    num_weight_bits: int
+    weight_exp: int
+
+
+def optimize_weight_bits(
+        weights: np.ndarray,
+        sign_mode: SignMode,
+        loihi2: ty.Optional[bool] = False) -> OptimizedWeights:
     """Optimizes the weight matrix to best fit in Loihi's synapse.
 
     Parameters
     ----------
-    weight : np.ndarray
-        Standard 8 bit signed weight matrix.
+    weights : np.ndarray
+        Standard 8-bit signed weight matrix.
+    sign_mode : SignMode
+        Determines whether the weights are purely excitatory, inhibitory,
+        or mixed sign.
     loihi2 : bool, optional
-        Flag to optimize for Lohi 2. By default False.
+        Flag to optimize for Loihi 2. By default False.
 
     Returns
     -------
-    np.ndarray
-        optimized weight matrix
-    int
-        weight bits
-    int
-        weight_exponent
-    SignMode
-        synapse sign mode
+    OptimizedWeights
+        An object that wraps the optimized weight matrix and weight parameters.
     """
-    max_weight = np.max(weight)
-    min_weight = np.min(weight)
+    if np.any(weights > 255) or np.any(weights < -256):
+        raise ValueError(f"weights have to be between -256 and 255. Got "
+                         f"weights between {np.min(weights)} and "
+                         f"{np.max(weights)}.")
 
-    if max_weight < 0:
-        sign_mode = SignMode.INHIBITORY
-        is_signed = 0
-    elif min_weight >= 0:
-        sign_mode = SignMode.EXCITATORY
-        is_signed = 0
-    else:
-        sign_mode = SignMode.MIXED
-        is_signed = 1
+    # Determine the exponent required to represent the weight matrix with a
+    # weight-matissa of up to 8 bits.
+    weight_exp = _determine_weight_exp(weights, sign_mode)
+    weights = np.left_shift(weights.astype(np.int32), -weight_exp)
+
+    num_weight_bits = _determine_num_weight_bits(weights, weight_exp, sign_mode)
+
+    if loihi2:
+        weights = weights // (1 << (8 - num_weight_bits))
+        if sign_mode == SignMode.MIXED:
+            weights = weights // 2
+
+    optimized_weights = OptimizedWeights(weights=weights.astype(int),
+                                         num_weight_bits=num_weight_bits,
+                                         weight_exp=weight_exp)
+    return optimized_weights
+
+
+def _determine_weight_exp(weights: np.ndarray,
```

Comment:
Please add a docstring

Reply:
Addressed in https://github.com/lava-nc/lava/pull/328/commits/7476abe0ebe64dd4fb79c079d553955a2f4f94fd